tests/integration/chroot.py

```diff
@@ -0,0 +1,298 @@
+          ssh_dir=chroot_ssh_dir))
+        cmd_ssh_keys.append('ssh-keygen -q -N "" -t ed25519 -f {ssh_dir}/ssh_host_ed25519_key'.format(
+          ssh_dir=chroot_ssh_dir))
+        for i in cmd_ssh_keys:
+            logger.info("Running command: {cmd}".format(
+              cmd=i))
+            p = subprocess.run([i], shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+            rc = p.returncode
+            if rc != 0:
+                logger.error("Command failed.")
+            else:
+                logger.info("Command sucessfully executed.")
+        # Generate chroot sshd directory for run/pid
+        chroot_sshd_run_dir = tmp_dir + "/run/sshd"
+        try:
+            os.mkdir(tmp_dir + "/run/sshd")
+            logger.info("Created directory: {dir}".format(
+              dir=chroot_sshd_run_dir))
+        except OSError:
+            logger.error("Could not create directory: {dir}".format(
+              dir=chroot_sshd_run_dir))
+        # Copy dedicated sshd config to chroot
+        sshd_config_src_file = "integration/misc/sshd_config_integration_tests"
+        chroot_sshd_cfg_dir = tmp_dir + "/etc/ssh/"
+        try:
+            shutil.copyfile(sshd_config_src_file, chroot_sshd_cfg_dir+"sshd_config_integration_tests")
+            logger.info("Copied sshd_config_integration_tests to: {dir}".format(
+              dir=chroot_sshd_cfg_dir))
+        except OSError:
+            logger.error("Could not copy sshd_config_integration_tests to {dir}".format(
+              dir=chroot_sshd_cfg_dir))
+        # Copy ssh / authorized_keys file for root user
+        chroot_root_dir = tmp_dir + "/root/"
+        chroot_ssh_authorized_keys = chroot_root_dir + ".ssh/authorized_keys"
+        ssh_authorized_keys = "/tmp/authorized_keys"
+        self._create_dir(chroot_root_dir+".ssh", 0o600)
+        try:
+            shutil.copyfile("/tmp/authorized_keys", chroot_ssh_authorized_keys)
+            logger.info("Copied authorized_keys to: {dir}".format(
+              dir=chroot_ssh_authorized_keys))
+        except OSError:
+            logger.error("Could not copy authorized_keys to {dir}".format(
+              dir=chroot_ssh_authorized_keys))
+
+
+    def _start_sshd_chroot(self, tmp_dir):
+        """ Start sshd inside the chroot """
+        # Define vars to have it more readable
+        gl_chroot_bin = "/gardenlinux/bin/garden-chroot"
+        chroot = tmp_dir
+        chroot_cmd = "/usr/sbin/sshd -D -f /etc/ssh/sshd_config_integration_tests"
+        # Execute in Popen as background task
+        # while we may perform our integration tests
+        proc_exec = "{chroot_bin} {chroot_env} {chroot_cmd}".format(
+          chroot_bin=gl_chroot_bin,
+          chroot_env=chroot,
+          chroot_cmd=chroot_cmd)
+        p = subprocess.Popen([proc_exec], shell=True)
+        logger.info("Started SSHD in chroot environment.")
+
+
+    def _wait_sshd(self):
```

Comment:
This method is used by `kvm` and `chroot`, no?
It may makes sense to move this method to another module. Maybe `sshclient.py`. What to you think?

Reply:
Ok, but I think we should raise a new issue for this - so we can adjust both ones directly. However, we may also adjust https://github.com/gardenlinux/gardenlinux/pull/712#discussion_r833204784.